### PR TITLE
Overhaul semantics of binary operators

### LINKED ISSUE #1103: Add more operators to shadow
We've currently only got `'#{+ - * / < <= = != >= >}` shadowed. Python has more operators than this, and so does Hy. I don't think there's any reason not to shadow these other binary operators: `'#{** @ // % << >> & ^ |}`.

The unary operator `'~` should also be easy to shadow, but the other unary operators `'#{+ -}` present some ambiguity. I propose importing the `operator.neg` and `operator.pos` functions for these, (as `neg` and `pos`).

There are a number of other operators in the `operator` module. The distinction between operator and function that Python has is unnatural in a Lisp like Hy. We should automatically import the remaining operators, perhaps with some renaming, as was done for `itertools`. We can treat these as a macro or special form when in the function position, but as a function name otherwise.

### LINKED ISSUE #1104: All shadowed operators should have the same arities as their counterparts.
Probably one of them got updated without the other:

``` Hy
=> (+)
0
=> ((lambda [op] (op)) +)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 1, in <lambda>
  File "c:\users\me\code\repos\hy/hy/core/shadow.hy", line 27, in +
    (defn + [&rest args]
  File "c:\users\me\code\repos\hy/hy/core/shadow.hy", line 31, in _hy_anon_fn_1
    (raise (TypeError "Need at least 1 argument to add/concatenate"))
TypeError: Need at least 1 argument to add/concatenate
```

We should check them all though.

### PR DESCRIPTION
**Pull request by Kodiologist**
Fixes #1103.

Fixes #1104.

### CODE REVIEW

File: hy/compiler.py [hunk 1]
```diff
@@ -1942,42 +1934,65 @@ def compile_maths_expression(self, expression):
         if PY35:
             ops.update({"@": ast.MatMult})
 
-        inv = expression.pop(0)
-        op = ops[inv]
+        op = ops[expression.pop(0)]
+        right_associative = op == ast.Pow
 
+        if right_associative:
+            expression = list(expression[::-1])
```
Comment (kirbyfan64) [inline hy/compiler.py]: Why is the outer `list` here necessary? `[::-1]` should already be returning a list anyway.

Comment (Kodiologist) [inline hy/compiler.py]: Fixed.

**Comment by Kodiologist** [comment]
@kirbyfan64, did you forget to approve this or have you simply not finished reading it?

**Pull request merged**